File: PlayerPRO 6/Plug-ins/Digital Plugs/Revert/PPRevertPlug.m
//
//  PPRevertPlug.m
//  PPMacho
//
//  Created by C.W. Betts on 9/6/14.
//
//

#import "PPRevertPlug.h"

@implementation PPRevertPlug

+ (BOOL)hasUIConfiguration
{
	return NO;
}

- (MADErr)runWithPcmd:(inout Pcmd*)myPcmd pluginInfo:(in PPInfoPlug *)thePPInfoPlug
{
	short	track, row;
	Pcmd	*srcCmd;
	long	memSize;
	
	memSize = myPcmd->structSize;
	
	srcCmd = (Pcmd*)calloc(memSize, 1);
	srcCmd->length = myPcmd->length;
	srcCmd->tracks = myPcmd->tracks;
	
	for (track = 0; track < myPcmd->tracks; track ++) {
		for (row = 0; row < myPcmd->length; row ++) {
			Cmd *myCmd, *myCmdsrc;
			myCmdsrc	= MADGetCmd(row, track, srcCmd);
			myCmd		= MADGetCmd(row, track, myPcmd);
			*myCmdsrc	= *myCmd;
		}
	}
	
	for (track = 0; track < myPcmd->tracks; track ++) {
		for (row = 0; row < myPcmd->length; row ++) {
			Cmd *myCmd, *myCmdsrc;
			myCmdsrc	= MADGetCmd(myPcmd->length - 1 - row, track, srcCmd);
			myCmd		= MADGetCmd(row, track, myPcmd);
			*myCmd		= *myCmdsrc;
		}
	}
	
	free(srcCmd);
	
	return MADNoErr;
}

@end

File: PlayerPRO 6/Plug-ins/Digital Plugs/Revert/PPRevertPlug.swift
//
//  File.swift
//  PPMacho
//
//  Created by C.W. Betts on 9/6/14.
//
//

import Foundation
import PlayerPROKit

public final class Revert: NSObject, PPDigitalPlugin {
	public let hasUIConfiguration = false
	
	public convenience init(forPlugIn: ()) {
		self.init()
	}
	
	override public init() {
		super.init()
	}

	public func runWithPcmd(myPcmd: UnsafeMutablePointer<Pcmd>, driver: PPDriver) -> MADErr {
		let srcCmd = PPKPcmd(myPcmd)
		let maxLenth = srcCmd.length - 1
		
		for track in 0 ..< srcCmd.tracks {
			for row in 0 ... maxLenth {
				let myCmdsrc = getCommand(row: maxLenth - row, track: track, aIntPcmd: srcCmd)
				replaceCommand(row: row, track: track, command: myCmdsrc, aPcmd: myPcmd)
			}
		}
		
		return .NoErr
	}
}
